Write public URL to marked base elements

siku2 opened the pull request:
Closes: #51

The second commit is the result of running `cargo fmt` (which defaults to the nightly version on my system). I included it here because these changes will come to Rustfmt stable sooner or later anyway.
Happy to remove if you don't like it!

I implemented this as a new step rather than including it in an existing one because I feel like it didn't fit anywhere else.
The implementation itself is almost disappointingly unsophisticated.
If you have any concerns or comments, there's no need to hold back :)

I would also like to update the "documentation" to mention this feature but I'm unsure where to place it. This isn't really an "asset" but perhaps it could be added to the "images & other resources" section which already references the `--public-url` flag?

thedodd reviewed (approved):
+1 Very nice! I think your decision on adding a new method for handling the finalization of the html was a good choice. 

Right now CI is setup to use the stable rustfmt ... so once you move this out of draft phase, it may give you an `x`. I am leaning towards keeping this project on the stable rustfmt for now. **Trust me,** there are lots of nightly and unstable lints that I want ... I think rustfmt has a lot of growing left to do. One of the lints I created (`space_around_attr_eq`). Anyway, I think it will be easier for many contributors to be able to stick with all stable toolchain.

Let me know what you think. Not a blocker.

thedodd commented:
Oh, @siku2 actually, do you mind just updating the changelog with some info on what was added here? You can keep it pretty brief. 

Then, as far as README update is concerned ... yea, we're accumulating a lot of new features. Going to have to create an mdBook github pages site soon. For now, perhaps you can add a `### directives` section after the `### images & other resources` section and add some details on how to use this new feature. Thoughts?

thedodd commented:
Also, I was wrong about rustfmt failing your PR. Looks like it is accepting the fmt changes. looks like they were just docstring related, so no worries about that. Keep 'em.

siku2 commented:
Alright, updated the `CHANGELOG` and added instructions to the `README`.
Btw, I would love to help out with the documentation.


> I think rustfmt has a lot of growing left to do.

Yeah, there are so many cases where I wish Rustfmt was more opinionated.
A lot of it is simply because the rules are still unstable (looking at you, `merge_imports`) but still.

thedodd commented on README.md:
```diff
@@ -104,6 +104,11 @@ This will allow your WASM application to reference images directly from the `dis
+You can instruct Trunk to write the URL passed to `--public-url` to the HTML output by adding this to your `<head>`: `<base data-trunk-public-url/>`.
+Trunk will set the `href` attribute of the element to the public URL. This changes the behavior of relative URLs to be relative to the public URL instead of the current location.
+You can also access this value at runtime using `document.baseURI` which is useful for apps that need to know what base URL they're hosted on (eg. for routing).
+
```
👌

thedodd commented:
@siku2 this is excellent! Thank you. Merging.
